plugins/experimental/log_requests/log_requests.cc

```diff
@@ -0,0 +1,352 @@
+{
+  std::vector<std::string> tokens;
+  size_t prev = 0, pos = 0;
+  do {
+    pos = str.find(delim, prev);
+    if (pos == std::string::npos)
+      pos = str.length();
+
+    std::string token = str.substr(prev, pos - prev);
+    if (!token.empty())
+      tokens.push_back(token);
+
+    prev = pos + delim.length();
+  } while (pos < str.length() && prev < str.length());
+
+  return tokens;
+}
+
+static bool
+should_log(TSHttpTxn txnp)
+{
+  TSMBuffer txn_resp_bufp;
+  TSMLoc txn_resp_loc;
+  TSHttpStatus resp_status;
+
+  if (TSHttpTxnClientRespGet(txnp, &txn_resp_bufp, &txn_resp_loc) != TS_SUCCESS) {
+    TSError(B_PLUGIN_NAME " Couldn't retrieve server response header.");
+    return false;
+  }
+
+  // get the transaction response status code
+  resp_status = TSHttpHdrStatusGet(txn_resp_bufp, txn_resp_loc);
+
+  // if resp_status is blacklisted, don't log
+  if (std::any_of(blacklist.begin(), blacklist.end(), [=](TSHttpStatus code) { return code == resp_status; }))
+    return false;
+
+  // else, if resp_status is in errors, log it
+  if (std::any_of(errors.begin(), errors.end(), [=](TSHttpStatus error) { return error == resp_status; }))
+    return true;
+
+  // fall through
+  return false;
+}
+
+static const std::string
+convert_http_version(const int version)
+{
+  if (version == TS_HTTP_VERSION(1, 0))
+    return "HTTP/1.0";
+  else if (version == TS_HTTP_VERSION(1, 1))
+    return "HTTP/1.1";
+  else if (version == TS_HTTP_VERSION(1, 2))
+    return "HTTP/1.2";
+  else if (version == TS_HTTP_VERSION(2, 0))
+    return "HTTP/2.0";
+  else {
+    return "(Unknown HTTP version)";
+  }
+}
+
+static void
+log_request_line(TSMBuffer bufp, TSMLoc loc, std::string output_header)
+{
+  int method_len;
+  int url_len;
+  const char *method;
+  const char *url;
+  TSMLoc url_loc;
+
+  // parse method
+  method = TSHttpHdrMethodGet(bufp, loc, &method_len);
+
+  // parse version
+  const std::string version = convert_http_version(TSHttpHdrVersionGet(bufp, loc));
+
+  // parse request line URL
+  TSHttpHdrUrlGet(bufp, loc, &url_loc);
+  url = TSUrlStringGet(bufp, url_loc, &url_len);
+
+  // get rid of the preceeding http:// on the request URI
+  for (int i = 0; i < 7; ++i)
+    url++;
+  url_len -= 7;
+
+  TSError(B_PLUGIN_NAME " [%s] request line is:\n%.*s %.*s %s\n", output_header.c_str(), method_len, method, url_len, url,
+          version.c_str());
+}
+
+static void
+log_response_status_line(TSMBuffer bufp, TSMLoc loc, std::string output_header)
+{
+  TSHttpStatus status_code;
+  const char *explanation;
+  int explanation_len;
+
+  // parse version
+  const std::string version = convert_http_version(TSHttpHdrVersionGet(bufp, loc));
+
+  // parse status code
+  status_code = TSHttpHdrStatusGet(bufp, loc);
+
+  // get explanation
+  explanation = TSHttpHdrReasonGet(bufp, loc, &explanation_len);
+
+  TSError(B_PLUGIN_NAME " [%s] response status line is:\n%s %d %.*s\n", output_header.c_str(), version.c_str(), status_code,
```

Comment:
I'm not sure I see the advantage. C++11 guarantees `c_str` to be a pointer to the internal `std::string` buffer anyways.